addon/model-data.js

```diff
@@ -17,10 +18,38 @@ function commitDataAndNotify(modelData, updates) {
 class M3SchemaInterface {
   constructor(modelData) {
     this.modelData = modelData;
+    this._keyBeingResolved = null;
+    this._refKeyUnknownKeysMap = {};
+  }
+
+  _beginUnknownKeyResolution(key) {
+    assert('_keyBeingResolved should be null.', this._keyBeingResolved === null);
+    this._keyBeingResolved = key;
+  }
+
+  _endUnknownKeyResolution(key) {
+    assert('_keyBeingResolved should be set.', key && this._keyBeingResolved === key);
+    this._keyBeingResolved = null;
+  }
+
+  _getDependentResolvedKeys(refKey) {
+    return this._refKeyUnknownKeysMap[refKey];
   }
 
   getAttr(name) {
-    return this.modelData.getAttr(name);
+    let value = this.modelData.getAttr(name);
+    const keyBeingResolved = this._keyBeingResolved;
+    assert('_keyBeingResolved should be set.', keyBeingResolved);
+
+    if (keyBeingResolved !== null && value) {
```

Comment:
Also guarding on truthiness of `value` isn't right.  We should always add `name` if it's being requested.

Imagine a flag whose value goes from `false` to `true` -- we still need to clear the cache under these conditions.

Reply:
Oh yeah good catch, that is the bad miss. I should check whether value is defined of not. I will update the code.